### DISCUSSION TITLE
When should I implement the methods body of IXunitSerializable (Deserialize,Serialize)?

### DISCUSSION THREAD

**Discussion by ruslanen** [xunit/xunit#2685] [Question] [answered]
Hi. I implemented interface `IXunitSerializable` for correctly running from Test Explorer, but didn't wrote body of `Deserialize` and `Serialize` methods because I don't understand when it needs.
```
    public void Deserialize(IXunitSerializationInfo info)
    {
    }

    public void Serialize(IXunitSerializationInfo info)
    {
    }
```
It doesn't make exceptions or break my tests. I put breakpoint to serialize and deserialize: `Deserialize` have never called, `Serialze` have called before tests running. I read it: https://github.com/xunit/xunit/discussions/2506 and still didn't understand.
Please, explain me what is scenario of using serialize/deserialize?

**Comment by bradwilson** [comment, marked as answer]
You are probably always doing "run all", which never deserializes (because it rediscovers the tests instead). If you were to try to run a single test (or a limited subset) via the Test Explorer UI, you will see Deserialize called.

You implement these to store and retrieve any custom data that your serialized object needs access to.

**Reply by ruslanen** [reply]
@bradwilson, thank you.
I run a single test but neither in `Serialize` nor `Deserialize` breakpoint called.

**Reply by bradwilson** [reply]
What kind of object are you serializing? Can you provide a repro project?

**Reply by ruslanen** [reply]
https://github.com/ruslanen/XUnitQuestion2685
@bradwilson it seems I understood my problem:
here I am using attribute which is calling `DeserializeObject` method instead of using native logic of deserializing in xUnit https://github.com/ruslanen/XUnitQuestion2685.sln/blob/master/XUnitQuestion2685.Tests/Tests.cs#L61
Am I right?

**Reply by bradwilson** [reply]
In this scenario, it's `TestData` which needs to be serializable, if you want to be able to display and run individual data rows from within Test Explorer.

I took your project and added a second data file (with failing data) to illustrate my point.

With your existing code, here's what Test Explorer looks like before tests are run:

![image](https://user-images.githubusercontent.com/16944/224158311-2a7aa504-74b3-40f7-8175-a8b3212284b5.png)

You'll note that there is just a single entry, even though there are two test files. Similarly, when I run them, we see one passing and one failing grouped into that single entry in the tree:

![image](https://user-images.githubusercontent.com/16944/224159049-e3fc5bec-37be-475a-85f9-afe1dc891412.png)

-----

Now here, I'll add serialization to the data row:

```diff
diff --git a/XUnitQuestion2685.Tests/TestData.cs b/XUnitQuestion2685.Tests/TestData.cs
index b463aa3..df6e4dd 100644
--- a/XUnitQuestion2685.Tests/TestData.cs
+++ b/XUnitQuestion2685.Tests/TestData.cs
@@ -1,9 +1,9 @@
 ﻿using System.Net;
+using Xunit.Abstractions;
 using XUnitQuestion2685.Api;

 namespace XUnitQuestion2685.Tests;

-public class TestData
+public class TestData : IXunitSerializable
 {
     public string TestDescription { get; set; } = "";

@@ -13,5 +13,21 @@ public class TestData

     public HttpStatusCode StatusCode { get; set; } = HttpStatusCode.OK;

+    public void Deserialize(IXunitSerializationInfo info)
+    {
+        TestDescription = info.GetValue<string>(nameof(TestDescription));
+        Query = info.GetValue<string>(nameof(Query));
+        Expected = info.GetValue<string[]>(nameof(Expected)).ToList();
+        StatusCode = info.GetValue<HttpStatusCode>(nameof(StatusCode));
+    }
+
+    public void Serialize(IXunitSerializationInfo info)
+    {
+        info.AddValue(nameof(TestDescription), TestDescription);
+        info.AddValue(nameof(Query), Query);
+        info.AddValue(nameof(Expected), Expected.ToArray());
+        info.AddValue(nameof(StatusCode), StatusCode);
+    }
+
     public override string ToString() => TestDescription;
 }
```

(Note that I serialized `Expected` as an array, since we support arrays but not the generic collection types.)

Now when I rebuild, each data row (in this case, each data file) is listed individually:

![image](https://user-images.githubusercontent.com/16944/224161437-7ae634b2-d8b5-4c18-8f31-5a25246fcc62.png)

This means you can run them individually as well:

![image](https://user-images.githubusercontent.com/16944/224161601-f76c4411-f83d-49f2-974f-e4ce2f033fca.png)

![image](https://user-images.githubusercontent.com/16944/224161762-cb9088d9-6e48-47e4-9aca-79d463275600.png)

**Reply by bradwilson** [reply]
It's important to note that your use of `JsonConvert.DeserializeObject` to parse the contents of the file is unrelated to the serialization we use when talking to Visual Studio. We don't care how the instances of `TestData` got populated; it's our job to serialize and deserialize them for Visual Studio once you've given them to us.

**Reply by ruslanen** [reply]
Oh, I got it.
I have tried to run tests in Visual Studio and got an error.
In JetBrains Rider it works without implementing Deserialize/Serialize (just empty methods).
Thank you very much, @bradwilson.
